Implement `workflow terminate` command (#436)
temporalio/cli: temporalcli/commandsmd/commands.md

* `--run-id`, `-r` (string) - Run Id. Cannot be set when query is set.
* `--query`, `-q` (string) - Start a batch to Signal Workflow Executions with given List Filter. Either this or
  Workflow Id must be set.

> * `--reason` (string) - Reason to perform batch. Only allowed if query is present unless the command specifies otherwise. Defaults to message with the current user's name.

Reviewer: This still mentions batch?

Author: You're looking at the help for `workflow signal` rather than `workflow terminate`. That command only allows a reason to be specified when invoked in batch mode. If you check the options for `workflow terminate` you'll see I updated it there already:

```
* `--reason` (string) - Reason to terminate this workflow. Defaults to message with the current user's name.
```

* `--run-id`, `-r` (string) - Run Id. Cannot be set when query is set.
* `--query`, `-q` (string) - Start a batch to Signal Workflow Executions with given List Filter. Either this or
  Workflow Id must be set.

> * `--reason` (string) - Reason to perform batch. Only allowed if query is present unless the command specifies otherwise. 

Reviewer 2: I think reason needs to be moved out of here and added individually for every command.
I tried phrasing it generically but it comes out unhelpful.

Author: Terminate seems to be the only command that needs it (from a brief grep of our protos), so I'm just going to break it out for this single command. All other batch ops can use this

Reviewer: schedule pause/unpause also takes a reason. It could be a different flag though.

### temporal workflow terminate: Terminate Workflow Execution by ID or List Filter.

The `temporal workflow terminate` command is used to terminate a [Workflow Execution](/concepts/what-is-a-workflow-execution). 
Canceling a running Workflow Execution records a `WorkflowExecutionTerminated` event as the closing Event in the workflow's Event History. 
Workflow code is oblivious to termination. Use `temporal workflow cancel` if you need to perform cleanup in your workflow.

Executions may be terminated by [ID](/concepts/what-is-a-workflow-id) with an optional reason:
```
temporal workflow terminate [--reason my-reason] --workflow-id MyWorkflowId
```

...or in bulk via a visibility query [list filter](/concepts/what-is-a-list-filter):
```
temporal workflow terminate --query=MyQuery
```

Use the options listed below to change the behavior of this command.

#### Options

* `--workflow-id`, `-w` (string) - Workflow Id. Either this or query must be set.
* `--run-id`, `-r` (string) - Run Id. Cannot be set when query is set.

> * `--query`, `-q` (string) - Start a batch to Signal Workflow Executions with given List Filter. Either this or Workflow Id must be set.

Reviewer 2: ```suggestion
* `--query`, `-q` (string) - Start a batch to Terminate Workflow Executions with given List Filter. Either this or Workflow Id must be set.
```

> * `--reason` (string) - Reason to terminate this workflow. Defaults to message with the current user's name.

Reviewer 2: Applies to both batch and single workflow:

```suggestion
* `--reason` (string) - Reason for termination. Defaults to message with the current user's name.
```

> * `--workflow-id`, `-w` (string) - Workflow Id. Either this or query must be set.
> * `--run-id`, `-r` (string) - Run Id. Cannot be set when query is set.
> * `--query`, `-q` (string) - Start a batch to terminate Workflow Executions with given List Filter. Either this or Workflow Id must be set.
> * `--reason` (string) - Reason for termination. Defaults to message with the current user's name.
> * `--yes`, `-y` (bool) - Confirm prompt to perform batch. Only allowed if query is present.

Reviewer 3: ```suggestion
Includes options set for [single workflow or batch](#options-set-for-single-workflow-or-batch).
```

Or are you making this a completely separate options set just because of the English description of reason?

Reviewer 2: Note that `#options-set-for-single-workflow-or-batch` mentions signal and reason is applied only to batch.

Author: Here's the discussion Roey and I had about this @cretz: https://github.com/temporalio/cli/pull/436#discussion_r1480582673

Reviewer 3: We can make docs more generic if we need to, but I have no problem copy/pasting just for docs-only reasons if we have to like here.

### temporal workflow terminate: Terminate Workflow Execution by ID or List Filter.

> The `temporal workflow terminate` command is used to terminate a [Workflow Execution](/concepts/what-is-a-workflow-execution). Canceling a running Workflow Execution records a `WorkflowExecutionTerminated` event as the closing Event in the workflow's Event History. No further command tasks may be scheduled after running this command.

Reviewer 2: ```suggestion
The `temporal workflow terminate` command is used to terminate a [Workflow Execution](/concepts/what-is-a-workflow-execution). Canceling a running Workflow Execution records a `WorkflowExecutionTerminated` event as the closing Event in the workflow's Event History. Workflow code is oblivious to termination. Use `temporal workflow cancel` if you need to perform cleanup in your workflow.
```

Executions may be terminated by [ID](/concepts/what-is-a-workflow-id):
```
temporal workflow terminate --workflow-id MyWorkflowId
```

...or in bulk via a visibility query [list filter](/concepts/what-is-a-list-filter):

```
temporal workflow terminate --query=MyQuery
```

Use the options listed below to change the behavior of this command.

#### Options

> Includes options set for [single workflow or batch](#options-set-single-workflow-or-batch)

Reviewer 2: The comment says `--reason` is only for batch (and mentions that it contains the time which is a lie), that should be fixed.
